### PR TITLE
在 Airbnb 使用机器学习预测房源的价值

### LINKED ISSUE #2009: 在 Airbnb 使用机器学习预测房源的价值
* 原文链接：[Using Machine Learning to Predict Value of Homes On Airbnb](https://medium.com/airbnb-engineering/using-machine-learning-to-predict-value-of-homes-on-airbnb-9272d3d4739d)
  * Markdown文件：[文件地址](https://github.com/xitu/gold-miner/blob/master/TODO/using-machine-learning-to-predict-value-of-homes-on-airbnb.md)
  * PR 地址：#2103
  * 文章分类：**AI**
  ----
  * 翻译时间：7 天
  * 校对时间：2 天
  * 翻译积分：7 分
  * 校对积分：2 分
  * [积分有什么用](https://github.com/xitu/gold-miner/wiki)

### PR DESCRIPTION
**Pull request by lsvih**
译文翻译完成，resolve #2009

### CODE REVIEW

**Comment by lsvih** [comment]
我觉得其实这是一篇招聘的软文。。。。

**Comment by TobiasLee** [comment]
@sqrthree  校对认领

**Comment by fanyijihua** [comment]
@TobiasLee 好的呢 :beer:

File: TODO/using-machine-learning-to-predict-value-of-homes-on-airbnb.md [hunk 1]
```diff
@@ -77,29 +77,30 @@ features: {
...
+- **数据插补**：我们需要检查是否有数据缺失，以及它是否为随机出现的缺失。如果不是随机现象，我们需要弄清楚其根本原因；如果是随机缺失，我们需要填充空缺数据。
+- **对分类进行编码**：通常来说我们不能在模型里直接使用原始的分类，因为模型并不能去拟合字符串。当分类数量比较少时，我们可以考虑使用 [one-hot encoding](http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.OneHotEncoder.html) 进行编码。如果分类数量比较多，我们就会考虑使用 [ordinal encoding](https://www.kaggle.com/general/16927), 按照分类的频率计数进行编码。
+
+在这一步中，我们还不知道最有效的一组特征时什么，因此需要我们能快速迭代代码。如 [Scikit-Learn](http://scikit-learn.org/stable/modules/generated/sklearn.pipeline.Pipeline.html)、[Spark](https://spark.apache.org/docs/latest/ml-pipeline.html) 等开源工具的 pipeline 结构对于原型构建来说是非常方便的工具。Pipeline 可以让数据科学家们设计蓝图，指定如何转换特征、如何训练模型。更具体来说，可以看下面我们 LTV 模型的 pipeline：
```
Comment (TobiasLee) [inline TODO/using-machine-learning-to-predict-value-of-homes-on-airbnb.md]: 最有效的一组特征时什么 -> 最有效的一组特征是什么

Comment (TobiasLee) [inline TODO/using-machine-learning-to-predict-value-of-homes-on-airbnb.md]: 如何训练模型 -> 训练哪一个模型

File: TODO/using-machine-learning-to-predict-value-of-homes-on-airbnb.md [hunk 2]
```diff
@@ -127,45 +128,47 @@ for field in categorical:
 features = FeatureUnion(transforms)
 ```
 
-At a high level, we use pipelines to specify data transformations for different types of features, depending on whether those features are of type binary, categorical, or numeric. [FeatureUnion](http://scikit-learn.org/stable/modules/generated/sklearn.pipeline.FeatureUnion.html) at the end simply combines the features column-wise to create the final training dataset.
+在高层设计时，我们使用 pipeline 来根据特征类型（如二进制特征、分类特征、数值特征等）来指定不同特征中数据的转换方式。最后使用 [FeatureUnion](http://scikit-learn.org/stable/modules/generated/sklearn.pipeline.FeatureUnion.html) 简单将特征列组合起来，形成最终的训练集。
 
-The advantage of writing prototypes with pipelines is that it abstracts away tedious data transformations using [data transforms](http://scikit-learn.org/stable/data_transforms.html). Collectively, these transforms ensure that data will be transformed consistently across training and scoring, which solves a common problem of data transformation inconsistency when translating a prototype into production.
+使用 pipeline 开发原型的优势在于，它可以使用 [data transforms](http://scikit-learn.org/stable/data_transforms.html) 来避免繁琐的数据类型转换。总的来说，这些转换是为了确保数据在训练和评估时保持一致，以避免将原型转成生产环境时出现的数据不一致的问题。
```
Comment (TobiasLee) [inline TODO/using-machine-learning-to-predict-value-of-homes-on-airbnb.md]: 以避免将原型转成生产环境时出现的数据不一致 ->
以避免将原型部署到生产环境时出现的数据不一致

File: TODO/using-machine-learning-to-predict-value-of-homes-on-airbnb.md [hunk 3]
```diff
@@ -197,33 +200,35 @@ def transform(X):
     return {'score': model['regressor'].predict(Xt)}
 ```
 
-Once the notebook is merged, ML Automator will wrap the trained model inside a [Python UDF](http://www.florianwilhelm.info/2016/10/python_udf_in_hive/) and create an [Airflow](https://airflow.incubator.apache.org/) pipeline like the one below. Data engineering tasks such as data serialization, scheduling of periodic re-training, and distributed scoring are all encapsulated as a part of this daily batch job. As a result, this framework significantly lowers the cost of model development for data scientists, as if there was a dedicated data engineer working alongside the data scientists to take the model into production!
+一旦 notebook 完成，ML Automator 将会把训练好的模型包装在 [Python UDF](http://www.florianwilhelm.info/2016/10/python_udf_in_hive/) 中，并创建一个 [Airflow](https://airflow.incubator.apache.org/) pipeline。数据序列化、定期重训练、分布式评价等数据工程任务都将被载入到日常批处理作业中。因此，这个框架显著降低了数据科学家将模型投入生产的成本，就像有一位数据工程师在与科学家一起工作一样！
```
Comment (TobiasLee) [inline TODO/using-machine-learning-to-predict-value-of-homes-on-airbnb.md]: 并创建一个 [Airflow] -> 并创建一个如下图所示的 [Airflow]

File: TODO/using-machine-learning-to-predict-value-of-homes-on-airbnb.md [hunk 4]
```diff
@@ -197,33 +200,35 @@ def transform(X):
...
 
 ![](https://cdn-images-1.medium.com/max/1600/1*DvPE_V_SoHV3pikOqiZxsg.png)
 
-A graph view of our LTV Airflow DAG, running in production
-**Note:***Beyond productionization, there are other topics, such as tracking model performance over time or leveraging elastic compute environment for modeling, which we will not cover in this post. Rest assured, these are all active areas under development.*
+我们 LTV 模型在 Airflow DAG 中的图形界面，运行与生产环境中
```
Comment (TobiasLee) [inline TODO/using-machine-learning-to-predict-value-of-homes-on-airbnb.md]: 运行与生产环境中 -> 运行于生产环境中

File: TODO/using-machine-learning-to-predict-value-of-homes-on-airbnb.md [hunk 5]
```diff
@@ -197,33 +200,35 @@ def transform(X):
...
-- **As a result, teams are more willing to invest in ML product ideas**: At the time of this post’s writing, we have several other teams exploring ML product ideas by following a similar approach: prioritizing the listing inspection queue, predicting the likelihood that listings will add cohosts, and automating flagging of low quality listings.
+过去的几个月中，我们的数据科学家们与 ML Infra 密切合作，产生了许多很好的模式和想法。我们相信这些工具将会为 Airbnb 开发机器学习模型开辟新的范例。
 
-We are very excited about the future of this framework and the new paradigm it brought along. By bridging the gap between prototyping and productionization, we can truly enable data scientists and engineers to pursue end-to-end machine learning projects and make our product better.
+- **首先，显著地降低了模型的开发成本**：通过组合各种不同的独立工具的优点（Zipline 用于特征工程、Pipeline 用于模型原型设计、AutoML 用于模型选择与验证，以及最后的 ML Automator 用于模型生产化），我们大大减短了模型的开发周期。
+- **其次，notebook 的设计减少了入门障碍**：还不熟悉框架的数据科学家可以立即得到大量的真实用例。在生产环境中，可以确保 notebook 是正确、自解释、最新的。这种设计模式收到了新用户的好评。
```
Comment (TobiasLee) [inline TODO/using-machine-learning-to-predict-value-of-homes-on-airbnb.md]: notebook 的设计减少了入门障碍 -> notebook 的设计降低了进入门槛

Comment (TobiasLee) [inline TODO/using-machine-learning-to-predict-value-of-homes-on-airbnb.md]: 这种设计模式收到了新用户的好评 -> 这种设计模式受到了新用户的好评

**Comment by RichardLeeH** [comment]
@sqrthree 校对认领

**Comment by fanyijihua** [comment]
@RichardLeeH 妥妥哒 :beers:

**Comment by lsvih** [comment]
@RichardLeeH 您好= =得麻烦您有空的时候校对一下哈

**Comment by RichardLeeH** [comment]
@lsvih 今天校对

**Comment by RichardLeeH** [comment]
@sqrthree @lsvih @TobiasLee 一校完毕

**Comment by lsvih** [comment]
@RichardLeeH 您没有提交 review 意见呀= =是不是没啥问题呢😜

**Comment by RichardLeeH** [comment]
@lsvih 我的失误，忘了点击提交了。现在应该能看到了

**Comment by lsvih** [comment]
@leviding 根据校对意见修改完毕

File: TODO/using-machine-learning-to-predict-value-of-homes-on-airbnb.md [hunk 6]
```diff
@@ -127,45 +128,47 @@ for field in categorical:
 features = FeatureUnion(transforms)
 ```
 
-At a high level, we use pipelines to specify data transformations for different types of features, depending on whether those features are of type binary, categorical, or numeric. [FeatureUnion](http://scikit-learn.org/stable/modules/generated/sklearn.pipeline.FeatureUnion.html) at the end simply combines the features column-wise to create the final training dataset.
+在高层设计时，我们使用 pipeline 来根据特征类型（如二进制特征、分类特征、数值特征等）来指定不同特征中数据的转换方式。最后使用 [FeatureUnion](http://scikit-learn.org/stable/modules/generated/sklearn.pipeline.FeatureUnion.html) 简单将特征列组合起来，形成最终的训练集。
 
-The advantage of writing prototypes with pipelines is that it abstracts away tedious data transformations using [data transforms](http://scikit-learn.org/stable/data_transforms.html). Collectively, these transforms ensure that data will be transformed consistently across training and scoring, which solves a common problem of data transformation inconsistency when translating a prototype into production.
+使用 pipeline 开发原型的优势在于，它可以使用 [data transforms](http://scikit-learn.org/stable/data_transforms.html) 来避免繁琐的数据类型转换。总的来说，这些转换是为了确保数据在训练和评估时保持一致，以避免将原型部署到生产环境时出现的数据不一致。
```
Comment (reid3290) [inline TODO/using-machine-learning-to-predict-value-of-homes-on-airbnb.md]: 避免繁琐的数据类型转换一句中的 data transformations 不如直接翻译成数据转换，因为“数据类型转换”好像有点多度解释了，而且还不一定对。

File: TODO/using-machine-learning-to-predict-value-of-homes-on-airbnb.md [hunk 7]
```diff
@@ -127,45 +128,47 @@ for field in categorical:
...
 
-Furthermore, pipelines also separates data transformations from model fitting. While not shown in the code above, data scientists can add a final step to specify an [estimator](http://scikit-learn.org/stable/tutorial/machine_learning_map/index.html) for model fitting. By exploring different estimators, data scientists can perform model selection to pick the best model to improve the model’s out of sample error.
+另外，pipeline 还可以将数据转换过程和训练模型过程分开。虽然上面代码中没有，但科学家可以在最后一步指定一种 [estimator（估值器）](http://scikit-learn.org/stable/tutorial/machine_learning_map/index.html)来训练模型。通过尝试使用不同的估值器，数据科学家可以为模型选出一个表现最佳的估值器，减少模型的样本误差。
```
Comment (reid3290) [inline TODO/using-machine-learning-to-predict-value-of-homes-on-airbnb.md]: 科学家 => 数据科学家，或者数据专家

File: TODO/using-machine-learning-to-predict-value-of-homes-on-airbnb.md [hunk 8]
```diff
@@ -127,45 +128,47 @@ for field in categorical:
...
+如上一节所述，我们需要确定候选模型中的哪个最适合投入生产。为了做这个决策，我们需要在模型的可解释性与复杂度中进行权衡。例如，稀疏线性模型的解释性很好，但它的复杂度太低了，不能很好地运作。一个足够复杂的树模型可以拟合各种非线性模式，但是它的解释性很差。这种情况也被称为[**偏差（Bias）和方差（Variance）的权衡**](http://scott.fortmann-roe.com/docs/BiasVariance.html)。
 
 ![](https://cdn-images-1.medium.com/max/1600/1*tQbBEq6T8ZJ9lFSCbZKFqw.png)
 
-Figure referenced from Introduction to Statistical Learning with R by James, Witten, Hastie, and Tibshirani
-In applications such as insurance or credit screening, a model needs to be interpretable because it’s important for the model to avoid inadvertently discriminating against certain customers. In applications such as image classification, however, it is much more important to have a performant classifier than an interpretable model.
+上图引用自 James、Witten、Hastie、Tibshirani 所著《R 语言统计学习》
```
Comment (reid3290) [inline TODO/using-machine-learning-to-predict-value-of-homes-on-airbnb.md]: image
格式好像有点问题

File: TODO/using-machine-learning-to-predict-value-of-homes-on-airbnb.md [hunk 9]
```diff
@@ -127,45 +128,47 @@ for field in categorical:
...
+由于模型的选择相当耗时，我们选择采用各种[自动机器学习](https://medium.com/airbnb-engineering/automated-machine-learning-a-paradigm-shift-that-accelerates-data-scientist-productivity-airbnb-f1f8a10d61f8)工具来加速这个步骤。通过探索大量的模型，我们最终会找到表现最好的模型。例如，我们发现 [XGBoost](https://github.com/dmlc/xgboost) (XGBoost) 明显比其他基准模型（比如 mean response 模型、岭回归模型、单一决策树）的表现要好。
 
 ![](https://cdn-images-1.medium.com/max/1600/1*y1O7nIxCFmgQamCfsrWfjA.png)
 
-Comparing RMSE allows us to perform model selection
-Given that our primary goal was to predict listing values, we felt comfortable productionizing our final model using XGBoost, which favors flexibility over interpretability.
+上图：通过比较 RMSE 可以让我们选择表现更好的模型
```
Comment (reid3290) [inline TODO/using-machine-learning-to-predict-value-of-homes-on-airbnb.md]: 通过...让... 
典型语病吧

File: TODO/using-machine-learning-to-predict-value-of-homes-on-airbnb.md [hunk 10]
```diff
@@ -127,45 +128,47 @@ for field in categorical:
...
+
+鉴于我们的最初目标是预测房源价格，因此我们很舒服地在最终的生产环境中使用 XGBoost 模型，比起可解释性它更注重于灵活性。
```
Comment (reid3290) [inline TODO/using-machine-learning-to-predict-value-of-homes-on-airbnb.md]: 灵活性？ 前面都是翻译成可解释性吧

File: TODO/using-machine-learning-to-predict-value-of-homes-on-airbnb.md [hunk 11]
```diff
@@ -77,29 +77,30 @@ features: {
...
+在构建训练集时，Zipline 将会找出训练集所需要的特征，使用智能秘钥连接后台，读取需要的数据并填充数据集。在构造房源 LTV 模型时，我们使用了一些 Zipline 中已经存在的特征，还自己写了一些特征。总体来说，模型中用了超过 150 个特征，其中包括：
 
-- **Location**: country, market, neighborhood and various geography features
-- **Price**: nightly rate, cleaning fees, price point relative to similar listings
-- **Availability**: Total nights available, % of nights manually blocked
-- **Bookability**: Number of bookings or nights booked in the past X days
-- **Quality**: Review scores, number of reviews, and amenities
+- **位置**：国家、市场、社区以及其它地理特征
+- **价格**：夜间费率、清洁费、与相似房源的价格差异
+- **可用性**：可在夜间预订的总天数，以及房主手动关闭夜间预订的占比百分数
```
[the remark this answers is not in the document]
Comment (lsvih) [reply TODO/using-machine-learning-to-predict-value-of-homes-on-airbnb.md]: 改成 “可否过夜” 了

File: TODO/using-machine-learning-to-predict-value-of-homes-on-airbnb.md (hunk 10 above)
Comment (lsvih) [reply TODO/using-machine-learning-to-predict-value-of-homes-on-airbnb.md]: 这儿是 flexibility，前面是说“性能”与“可解释性”做取舍。在这儿我改成“模型的弹性”了。

**Comment by lsvih** [comment]
@reid3290 修改了

**Comment by leviding** [comment]
@lsvih  已经 merge 啦~ 快快麻溜发布到掘金专栏然后给我发下链接，方便及时添加积分哟。

掘金翻译计划有自己的知乎专栏，你也可以投稿哈，推荐使用一个[好用的插件](https://chrome.google.com/webstore/detail/minimalist-markdown-edito/pghodfjepegmciihfhdipmimghiakcjf?hl=zh-CN)。
专栏地址：https://zhuanlan.zhihu.com/juejinfanyi

**Comment by lsvih** [comment]
@leviding 已经发好了，https://juejin.im/post/59acfc336fb9a0249471e47d
